auto_round/modelling/replace_modules.py

```diff
@@ -0,0 +1,180 @@
+                raise ValueError(
+                    f"Module '{cls.original_module_class()}' already registered to "
+                    f"{existing.__name__}. Cannot register {cls.__name__}."
+                )
+
+            cls._replacement_registry[cls.original_module_class()] = cls
+            logger.trace(f"Registered {cls.__name__} for replacing {cls.original_module_class()}")
+
+    @classmethod
+    def get_replacement_class(cls, module_class_name: str) -> Type["ReplacementModuleBase"]:
+        """Get replacement class for a given module class name."""
+        return cls._replacement_registry.get(module_class_name)
+
+    @classmethod
+    def is_registered(cls, module_class_name: str) -> bool:
+        """Check if a module class has a replacement implementation."""
+        return module_class_name in cls._replacement_registry
+
+    @classmethod
+    def is_to_be_replaced(
+        cls,
+        original: torch.nn.Module,
+        module_class_name: str,
+    ) -> bool:
+        """Determine if the given module should be replaced.
+
+        Users can extend this method to add custom logic for replacement.
+        """
+        return cls.is_registered(module_class_name)
+
+    @classmethod
+    def get_registered_modules(cls) -> list:
+        """Get list of all registered module class names."""
+        return list(cls._replacement_registry.keys())
+
+    @classmethod
+    @abstractmethod
+    def original_module_class(cls) -> str:
+        """Return the class name of the module this replaces."""
+        pass
+
+    @classmethod
+    @abstractmethod
+    def from_original(
+        cls,
+        original: torch.nn.Module,
+        config,
+    ) -> "ReplacementModuleBase":
+        """Create replacement module from original module."""
+        pass
+
+
+# Note: adapted from llm-compressor
+# https://github.com/vllm-project/llm-compressor/blob/main/src/llmcompressor/modeling/moe_context.py
+
+
+def apply_replacements(
+    model: torch.nn.Module,
+) -> torch.nn.Module:
+    """
+    Function to apply module replacements to a model.
+
+    This scans all modules in the model and replaces any registered modules with their
+    replacement equivalents. Non-permanent modules are tracked for later restoration.
+
+    The model is modified in-place, so the same model object should be used.
+
+    Args:
+        model: The model to apply module replacement to (modified in-place).
+
+    Returns:
+        The model with modules replaced.
+    """
+    _import_required_replacements(model)
+    replaced = {}
+
+    # Step 1: Collect all modules that need replacement
+    logger.debug("Scanning for modules to replace")
+    modules_to_replace = []
+    for name, module in model.named_modules():
+        class_name = (
+            module.original_module_class()
+            if hasattr(module, "original_module_class") and callable(module.original_module_class)
```

Comment:
if the module has attrbute original_module_class, the module is already replaced. Why do we need to replace it again?